src/vm/costs/mod.rs

```diff
@@ -134,6 +211,73 @@ impl LimitedCostTracker {
     }
 }
 
+fn parse_cost(cost_function: ClarityCostFunction, eval_result: InterpreterResult<Option<Value>>) -> Result<ExecutionCost> {
+    match eval_result {
+        Ok(Some(Value::Tuple(data))) => {
+            let results = (
+                data.data_map.get("write_length"),
+                data.data_map.get("write_count"),
+                data.data_map.get("runtime"),
+                data.data_map.get("read_length"),
+                data.data_map.get("read_count"),
+            );
+
+            match results {
+                (Some(UInt(write_length)),
+                    Some(UInt(write_count)),
+                    Some(UInt(runtime)),
+                    Some(UInt(read_length)),
+                    Some(UInt(read_count))) => {
+                    Ok(ExecutionCost {
+                        write_length: (*write_length as u64),
+                        write_count: (*write_count as u64),
+                        runtime: (*runtime as u64),
+                        read_length: (*read_length as u64),
+                        read_count: (*read_count as u64),
+                    })
+                },
+                _ => Err(CostErrors::CostComputationFailed("Execution Cost tuple does not contain only UInts".to_string()))
+            }
+        },
+        Ok(Some(_)) => Err(CostErrors::CostComputationFailed("Clarity cost function returned something other than a Cost tuple".to_string())),
+        Ok(None) => Err(CostErrors::CostComputationFailed("Clarity cost function returned nothing".to_string())),
+        Err(e) => Err(CostErrors::CostComputationFailed(format!("Error evaluating result of cost function {}: {}", cost_function.get_name(), e))),
+    }
+}
+
+fn compute_cost(
+    cost_tracker: &mut LimitedCostTracker,
+    cost_function: ClarityCostFunction,
+    input_size: u64) -> Result<ExecutionCost> {
```

Comment:
I think this method does a lot of unnecessary duplicate computation -- it re-parses and re-evals the same cost functions over and over again.  I think we talked about making it so the node parses the cost functions once per block, and caches them over the course of evaluating the block so it doesn't have to do this?  You could put this cache into a `ClarityBlockConnection`, for example, and have it be instantiated as part of its constructor.

Reply:
This method uses pre-parsed ASTs to avoid the overhead of re-parsing, and the LimitedCostTracker has the same lifetime as a block. That combined with the fact that the cost tracker is also what is invoking this function, I think it makes sense to keep the cache in the tracker itself (as is currently done). I do think there's still a lot of opportunity for speedups here, but I would note that this PR is opened against `next-cost` with the intention being that there are going to be a lot of subsequent PRs both improving the efficiency of the basic implementation here, as well as implementing the remaining items to support cost voting.